modules/unspents/src/dimensions.ts

```diff
@@ -1,287 +1,151 @@
-  nP2trScriptPathLevel2Inputs: number;
-  // Single-Signature
-  nP2shP2pkInputs: number;
-  outputs: IOutputDimensions;
-}
+export class OutputDimensions {
+  /**
+   * Number of outputs
+   */
+  count: number;
+  /**
+   * Aggregate vSize
+   */
+  size: number;
 
-export interface IDimensions extends IBaseDimensions {
-  nInputs: number;
-  nOutputs: number;
+  constructor({ count = 0, size = 0 }: OutputDimensions = { count: 0, size: 0 }) {
+    if (count === 0 || size === 0) {
+      if (count !== 0 || size !== 0) {
+        throw new Error(`count and size must both be zero if one is zero`);
+      }
+    }
 
-  plus(v: Partial<IDimensions>): IDimensions;
-  times(n: number): IDimensions;
+    this.count = count;
+    this.size = size;
 
-  isSegwit(): boolean;
-  getNInputs(): number;
-  getOverheadVSize(): number;
-  getInputsVSize(): number;
-  getOutputsVSize(): number;
-  getVSize(): number;
+    Object.freeze(this);
+  }
 }
 
-interface IFromInputParams {
+interface FromInputParams {
   // In cases where the input type is ambiguous, we must provide a hint about spend script type.
   assumeUnsigned?: symbol;
 }
 
-export interface IFromUnspentParams {
+export interface FromUnspentParams {
   // The p2tr output type has multiple spend options and thus different weights per spend path.
   p2trSpendType: 'keypath' | 'scriptpath-level1' | 'scriptpath-level2';
 }
 
-export interface IDimensionsStruct extends t.Struct<IDimensions> {
-  (v: IBaseDimensions): IDimensions;
-  ASSUME_P2SH: symbol;
-  ASSUME_P2SH_P2WSH: symbol;
-  ASSUME_P2WSH: symbol;
-  ASSUME_P2TR_KEYPATH: symbol;
-  ASSUME_P2TR_SCRIPTPATH_LEVEL1: symbol;
-  ASSUME_P2TR_SCRIPTPATH_LEVEL2: symbol;
-
-  SingleOutput: {
-    p2sh: IDimensions;
-    p2shP2wsh: IDimensions;
-    p2wsh: IDimensions;
-    p2tr: IDimensions;
-    p2pkh: IDimensions;
-    p2wpkh: IDimensions;
-  };
-  new (v: IBaseDimensions): IDimensions;
+/**
+ * Dimensions of a BitGo wallet transactions.
+ */
+export class Dimensions {
+  static ASSUME_P2SH = Symbol('assume-p2sh');
+  static ASSUME_P2SH_P2WSH = Symbol('assume-p2sh-p2wsh');
+  static ASSUME_P2WSH = Symbol('assume-p2wsh');
+  static ASSUME_P2TR_KEYPATH = Symbol('assume-p2tr-keypath');
+  static ASSUME_P2TR_SCRIPTPATH_LEVEL1 = Symbol('assume-p2tr-scriptpath-level1');
+  static ASSUME_P2TR_SCRIPTPATH_LEVEL2 = Symbol('assume-p2tr-scriptpath-level2');
+
+  /** Input counts for BitGo wallet multi-signature inputs */
+  public nP2shInputs = 0;
+  public nP2shP2wshInputs = 0;
+  public nP2wshInputs = 0;
+  public nP2trKeypathInputs = 0;
+  public nP2trScriptPathLevel1Inputs = 0;
+  public nP2trScriptPathLevel2Inputs = 0;
+
+  /* Input count for single-signature inputs (Replay Protection inputs) */
+  public nP2shP2pkInputs = 0;
+
+  public outputs: OutputDimensions = new OutputDimensions();
+
+  constructor(d: Partial<Dimensions> = {}) {
+    Object.entries(d).forEach(([key, value]) => this.setProperty(key, value));
+
+    Object.freeze(this);
```

Comment:
Does it make sense to set all the props above to `readonly` instead of use `freeze`? That way we ensure no props are changed at compile time rather than runtime I believe.

If that's done you might need to move `setProperty` inside the constructor but that's fine since we don't want to call it anywhere besides the constructor anyway.

Reply:
here is what happened: just changing 

```
  public nP2shInputs = 0;
```

to

```
  public readonly nP2shInputs = 0;
```

gives a build error (that I didn't look at too closely). TS infers that the type is `0` (not `number`!)

with
```

```
  public readonly nP2shInputs: number = 0;
```
it works fine

Reply:
I will keep `freeze()` for runtime safety in potential JS callsites

Reply:
Interesting. I think the approach of `:number = 0` is fine, another option is to deconstruct the parameters to the constructor and set defaults there, then assign line by line in the constructor rather than via the `setParameter` loop, as in:

```typescript
public readonly nP2shInputs: number;
constructor({ nP2shInputs = 0... }: { nP2shInputs?: number...}) {
  this.nP2shInputs = nP2shInputs;
  ..
```

However this would be more verbose so fine to leave as is, just something to consider.